File: app/Quotes/bs/src/widgets/wserver_id.ml
(*  Copyright 30-Jul-2018 ºDeme
    GNU General Public License - V3 <http://www.gnu.org/licenses/>
*)

open Ui
open I18n

type t = {
  server: string;
  nick : string;
  icon : Domo.t;
  msg : Domo.t;
  field : Domo.t;
  w : Domo.t
}

let send rq f = Client.send (Main.client ()) rq f

let send_async rq f = Client.send_async (Main.client ()) rq f

let wrq = Client.wrq

let rrp = Client.rrp

let server field = field.server

let nick field = field.nick

let field f = f.field

let widget field = field.w

let download field = Json.(
  let {server; icon; msg; field; _} = field in
  let value = Txt.trim (Domo.value field) in
  let t = Js.Date.make () in
  if value = "" then alert (i18 "Field is empty")
  else (
    ignore Domo.(
      remove_all icon |> add [q "img" [Att ("src", "img/wait.gif")][]]);
    let rq = wrq [
      "page", wstring "wserverId";
      "rq", wstring "test";
      "server", wstring server;
      "code", wstring value
    ] in
    send_async rq (fun rp -> (
      let ok = rrp rp "ok" rbool in
      if ok
      then (
        let t2 = Js.Date.make () in
        let t = Dec.to_str 2 (Js.Date.(getTime t2 -. getTime t) /. 1000.)
        in (
          ignore Domo.(remove_all icon |> add [img "well"]);
          ignore Domo.(msg |>
            set [Html (i18f (i18 " (%0 seconds)") [t])])))
      else (
        ignore Domo.(remove_all icon |> add [img "error"]);
        ignore Domo.(msg |> set [Html ""]))))))

let change nick server value =
  let value = Txt.trim value in Json.(
    if value = "" then ()
    else
      let rq = wrq [
        "page", wstring "wserverId";
        "rq", wstring "setCode";
        "server", wstring server;
        "nick", wstring nick;
        "code", wstring value
      ] in
      send rq (fun _ -> ())
  )

let mk nick server =
  let icon = q "div" [][img "unknown"]
  and msg = q "div" [Style "text-align:left;"][]
  and field = q "input" [
      Style "width:150px"
    ][]
  and download_td = q "td" [Style "width:5px"][]
  in
  let field = field |>
    Domo.(set [On ("change", (fun _ -> change nick server (value field)))])
  in
  let w = q "table" [Att ("align", "center")][
      q "tr" [][
        download_td;
        q "td" [Style "width:5px"][icon];
        q "td" [][msg]];
      q "tr" [][q "td" [Att_i ("colspan", 3)][field]]]
  in
  let r = {server; nick; icon; msg; field; w}
  in Json.(
    let rq = wrq [
      "page", wstring "wserverId";
      "rq", wstring "getCode";
      "server", wstring server;
      "nick", wstring nick
    ] in
    send rq (fun rp ->
      match rrp rp "code" (ropt rstring) with
      | None -> ignore Domo.(remove_all icon |> add [img "error"])
      | Some code -> ignore (field |> Domo.set [Value code])
    );
    ignore (download_td |> Domo.add [
            link (fun _ -> download r) |>
              Domo.set [Att ("title", "Test")] |>
              Domo.add [img "download"]]);
    r
  )

File: app/Quotes/js/src/wgs/WserverId.js
// Copyright 13-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// eslint-disable-next-line
import Main from "../Main.js";
import Ui from "../dmjs/Ui.js";
// eslint-disable-next-line
import Domo from "../dmjs/Domo.js";
import Dec from "../dmjs/Dec.js";
import {_} from "../I18n.js";

const $ = Ui.$;

/** Widget for reading a server identifier */
export default class WserverId {
  /**
   * @param {!Main} main
   * @param {string} serverId
   * @param {string} nickId
   * @param {string} nickCode Server code.
   */
  constructor (main, serverId, nickId, nickCode) {
    this._main = main;
    this._serverId = serverId;
    this._nickId = nickId;
    this._nickCode = nickCode;

    this._icon = $("td").style("width:5px;");
    this._msg = $("span");
    this._field = $("input").style("width:150px");
  }

  /** @return {string} */
  get nickId () {
    return this._nickId;
  }

  /** @return {string} */
  get serverId () {
    return this._serverId;
  }

  /** @return {string} */
  get nickCode () {
    return this._nickCode;
  }

  async test () {
    const code = this._field.value().trim();
    this._icon.removeAll().add($("img").att("src", "img/wait.gif"));
    this._msg.removeAll();
    if (code === "") {
      this._icon.removeAll().add(Ui.img("error"));
      return;
    }
    const now = Date.now();
    const data = {
      "source": "wserverId",
      "rq": "testCode",
      "serverId": this._serverId,
      "code": code
    };
    const rp = await this._main.client.send(data);
    const e = rp["error"];
    if (e === "") {
      const dif = new Dec((Date.now() - now) / 1000, 3);
      this._icon.removeAll().add(Ui.img("well"));
      this._msg.removeAll().text(`${dif.toEu()} secs.`);
    } else {
      this._icon.removeAll().add(Ui.img("error").att("title", e));
    }
  }

  change () {
    const code = this._field.value().trim();
    if (code === "") {
      return;
    }
    const data = {
      "source": "wserverId",
      "rq": "setCode",
      "serverId": this._serverId,
      "nickId": this._nickId,
      "code": code
    };
    this._main.client.send(data);
  }

  // ____
  // View ------------------------------------------------------------
  // TTTT

  /** @return {!Domo} */
  wg () {
    const self = this;

    self._field.value(self._nickCode).on("change", () => {
      self.change();
    });
    return $("table").att("align", "center")
      .add($("tr")
        .add($("td").style("width:5px;").add(Ui.link(() => {
          self.test();
        }).add(Ui.img("download")).att("title", _("Test"))))
        .add(self._icon.add(Ui.img("unknown")))
        .add($("td").add(self._msg)))
      .add($("tr")
        .add($("td").att("colspan", 3).add(self._field)))
    ;

  }
}
